openjdk/mmtkHeap.cpp

```diff
@@ -90,6 +90,10 @@ jint MMTkHeap::initialize() {
     bool set_options = process_bulk(strdup(ThirdPartyHeapOptions));
     guarantee(set_options, "Failed to set MMTk options. Please check if the options are valid: %s\n", ThirdPartyHeapOptions);
   }
+
+  // Read MMTk options from environment variables (such as `MMTK_THREADS`), overriding cmdline options.
+  mmtk_builder_read_env_var_settings();
```

Comment:
The bug in #244 is that the env var value got overwritten by the command line value. But with the change, the command line value will get overwritten by the env var value. So the PR does not solve the issue, it just changes the priority of the option value from different sources.

We should not mix the use of env vars and command line options. Maybe it is time to stop using env vars. It may take some time for the scripts and running configs to keep up with the change. I guess we could make it a goal for the next release. We would have 6 weeks to fix broken stuff.

https://github.com/mmtk/mmtk-core/pull/955 is good. It allows us to disable setting options by env var easily. We could also accept https://github.com/mmtk/mmtk-core/pull/954. It is a workaround for the immediate issue #244. In the long run, it is also okay to allow users to check if an option is explicitly set or not.

Reply:
In short, my suggestion is we review both mmtk-core PRs, and we move away from env var in the next release.

Reply:
Actually I believe this solves the issue. For openjdk, there are three ways to set options: (1) `MMTK_` env vars (2) ThirdPartyHeapOptions (3) Copied from hotspot cmd args

The problem is that we're ignoring env var values and unconditionally using openjdk cmd arg for the `threads` option. This should be solved as long as we define a clear priority between these three sources, and make sure the one with higher priority can overwrite others, instead of ignoring it.

Reply:
Tbh I'm not sure why you can't set `ParallelGCThreads` for OpenJDK instead of `MMTK_THREADS` in #242. Is there a particular reason we need to support that?

Reply:
> Actually I believe this solves the issue. For openjdk, there are three ways to set options: (1) `MMTK_` env vars (2) ThirdPartyHeapOptions (3) Copied from hotspot cmd args
> 
> The problem is that we're ignoring env var values and unconditionally using openjdk cmd arg for the `threads` option. This should be solved as long as we define a clear priority between these three sources, and make sure the one with higher priority can overwrite others, instead of ignoring it.

OK. I guess the actual issue is that command line args always have a default value, so when we prioritise command line args, it always overwrites values from other sources. But if we prioritise env vars, it could be just empty if not set. If you think https://github.com/mmtk/mmtk-core/pull/954 and the OpenJDK PR is not needed, you can close them.

Reply:
> Tbh I'm not sure why you can't set ParallelGCThreads for OpenJDK instead of MMTK_THREADS in https://github.com/mmtk/mmtk-openjdk/issues/242. Is there a particular reason we need to support that?

`ParallelGCThreads` should work. The problem is we can only set `ParallelGCThreads` now. The binding will ignore the values in `MMTK_THREADS`

Reply:
> OK. I guess the actual issue is that command line args always have a default value, so when we prioritise command line args, it always overwrites values from other sources. But if we prioritise env vars, it could be just empty if not set.

Yes.  That's the cause.  But OpenJDK actually provides a way to tell whether a command line option is explicitly given or omitted as default (`FLAG_IS_DEFAULT(ParallelGCThreads)`).  I have adjusted the PR so that environment variables overrides the *default* of `ParallelGCThreads`, but if `-XX:ParallelGCThreads=xxxx` is given on the command line, it will override the environment variable instead.

> If you think [mmtk/mmtk-core#954](https://github.com/mmtk/mmtk-core/pull/954) and the OpenJDK PR is not needed, you can close them.

But after a second thought, https://github.com/mmtk/mmtk-core/pull/954 may be useful if we want to optionally pass non-default MMTk options to other places just like we optionally pass non-default OpenJDK options to MMTk.  We can keep both https://github.com/mmtk/mmtk-core/pull/954 and https://github.com/mmtk/mmtk-core/pull/955.

Reply:
> > Tbh I'm not sure why you can't set ParallelGCThreads for OpenJDK instead of MMTK_THREADS in #242. Is there a particular reason we need to support that?
> 
> `ParallelGCThreads` should work. The problem is we can only set `ParallelGCThreads` now. The binding will ignore the values in `MMTK_THREADS`

That's what I'm saying, I'm not sure what the problem is with using `ParallelGCThreads`? Shouldn't we try to be as similar to the VM as possible? `MMTK_THREADS` and other environment variables are (still) there to quickly bootstrap new bindings/ports so that they don't have to bother fiddling with parsing commandline options.